roscala/src/main/scala/coop/rchain/roscala/ob/Ob.scala

```diff
@@ -1,20 +1,40 @@
 package coop.rchain.roscala.ob
 
+import com.typesafe.scalalogging.Logger
+
 import scala.collection.mutable
+import Ob.logger
+import coop.rchain.roscala.GlobalEnv
+import coop.rchain.roscala.Vm.State
 
 abstract class Ob {
   val slot: mutable.ArrayBuffer[Ob] = mutable.ArrayBuffer[Ob]()
   var meta: Meta                    = _
   var parent: Ob                    = _
 
-  def dispatch(ctxt: Ctxt): Ob = Niv
+  def dispatch(state: State, globalEnv: GlobalEnv): Ob = Niv
+
+  def invoke(state: State, globalEnv: GlobalEnv): Ob = Niv
 
-  def invoke(ctxt: Ctxt): Ob = Niv
+  /** Tries to lookup value for key and then invokes the value
+    *
+    * `ctxt.trgt` contains the key.
+    */
+  def lookupAndInvoke(state: State, globalEnv: GlobalEnv): Ob = {
+    val fn = meta.lookupObo(this, state.ctxt.trgt, state.ctxt)(globalEnv)
+    logger.debug(s"Lookup and invoke $fn")
+    fn.invoke(state, globalEnv)
+  }
 
-  def lookupAndInvoke(ctxt: Ctxt): Ob = {
-    val fn = meta.lookupObo(this, ctxt.trgt, ctxt)
+  def lookup(key: Ob)(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+    logger.debug(s"Lookup for $key in $this")
+    val me     = this
+    val result = meta.get(me, key)(globalEnv)
 
-    fn.invoke(ctxt)
+    if (result == Absent)
+      parent.lookup(key)(ctxt, globalEnv)
```

Comment:
What is the purpose to pass `ctxt` here?

Reply:
Good observation.